pkg/ingester/circuitbreaker.go

```diff
@@ -0,0 +1,233 @@
+		circuitBreakerCurrentStateGaugeFn(s)
+		// We initialize all possible states for the circuitBreakerTransitions metrics
+		cbMetrics.circuitBreakerTransitions.WithLabelValues(s.String())
+	}
+	for _, r := range []string{resultSuccess, resultError, resultOpen} {
+		// We initialize all possible results for the circuitBreakerResults metrics
+		cbMetrics.circuitBreakerResults.WithLabelValues(r)
+	}
+	return cbMetrics
+}
+
+type CircuitBreakerConfig struct {
+	Enabled                    bool          `yaml:"enabled" category:"experimental"`
+	FailureThresholdPercentage uint          `yaml:"failure_threshold_percentage" category:"experimental"`
+	FailureExecutionThreshold  uint          `yaml:"failure_execution_threshold" category:"experimental"`
+	ThresholdingPeriod         time.Duration `yaml:"thresholding_period" category:"experimental"`
+	CooldownPeriod             time.Duration `yaml:"cooldown_period" category:"experimental"`
+	InitialDelay               time.Duration `yaml:"initial_delay" category:"experimental"`
+	PushTimeout                time.Duration `yaml:"push_timeout" category:"experiment"`
+	testModeEnabled            bool          `yaml:"-"`
+}
+
+func (cfg *CircuitBreakerConfig) RegisterFlags(f *flag.FlagSet) {
+	prefix := "ingester.circuit-breaker."
+	f.BoolVar(&cfg.Enabled, prefix+"enabled", false, "Enable circuit breaking when making requests to ingesters")
+	f.UintVar(&cfg.FailureThresholdPercentage, prefix+"failure-threshold-percentage", 10, "Max percentage of requests that can fail over period before the circuit breaker opens")
+	f.UintVar(&cfg.FailureExecutionThreshold, prefix+"failure-execution-threshold", 100, "How many requests must have been executed in period for the circuit breaker to be eligible to open for the rate of failures")
+	f.DurationVar(&cfg.ThresholdingPeriod, prefix+"thresholding-period", time.Minute, "Moving window of time that the percentage of failed requests is computed over")
+	f.DurationVar(&cfg.CooldownPeriod, prefix+"cooldown-period", 10*time.Second, "How long the circuit breaker will stay in the open state before allowing some requests")
+	f.DurationVar(&cfg.InitialDelay, prefix+"initial-delay", 0, "How long the circuit breaker should wait between creation and starting up. During that time both failures and successes will not be counted.")
+	f.DurationVar(&cfg.PushTimeout, prefix+"push-timeout", defaultPushTimeout, "How long is execution of ingester's Push supposed to last before it is reported as timeout in a circuit breaker. This configuration is used for circuit breakers only, and timeout expirations are not reported as errors")
+}
+
+type circuitBreaker struct {
+	cfg     CircuitBreakerConfig
+	logger  log.Logger
+	metrics *circuitBreakerMetrics
+	active  atomic.Bool
+	cb      circuitbreaker.CircuitBreaker[any]
+}
+
+func newCircuitBreaker(cfg CircuitBreakerConfig, isActive bool, logger log.Logger, registerer prometheus.Registerer) *circuitBreaker {
```

Comment:
Do we need `isActive`, or can we always create CB as inactive?

Reply:
Our default configuration is with active circuit breakers, so IMO it makes no sense to create them as inactive by default and then manually activate them. I think it makes sense to specify in the constructor if we want the circuit breaker to be created as active or inactive. Moreover, it simplifies tests

Reply:
> Our default configuration is with active circuit breakers

I don't understand. Ingesters always create inactive circuit breakers, right?